src/library/scala/collection/immutable/Stream.scala

```diff
@@ -1293,6 +1244,23 @@ object Stream extends SeqFactory[Stream] {
   private[immutable] def collectedTail[A, B, That](head: B, stream: Stream[A], pf: PartialFunction[A, B], bf: CanBuildFrom[Stream[A], B, That]) = {
     cons(head, stream.tail.collect(pf)(bf).asInstanceOf[Stream[B]])
   }
-}
-
 
+  /** An implementation of `FilterMonadic` allowing GC of the `Stream` as it is processed.
+    *
+    * Because this is not an inner class of `Stream`, it is now possible to offer explicit
+    * 'null after first use' semantics for the reference to the head of the original
+    * `Stream`, allowing GC to collect while the tail is still processing (see SI-8990).
+    */
```

Comment:
You can't avoid the head reference, but in obeying the _spirit_ rather than the letter of the API, you _can_ discard all of the leading and filtered-out elements.  For non-memoizing eager collections, restricting the domain is the thing to do to save memory: you've already got it all, just pick the ones you want.  But for memoizing lazy collections, you want _exactly_ the opposite.  The only difference is that you'd like to throw away the head lazily also, which regular `filter` doesn't support.  So `withFilter`'s job should be to take the original collection, call `filter` on it at most one time, and store that result instead.  That way the intent of `withFilter` is filled: do as little work as possible.